pinot-spi/src/main/java/org/apache/pinot/spi/data/readers/PrimaryKey.java

```diff
@@ -37,7 +41,57 @@ public Object[] getValues() {
   }
 
   public byte[] asBytes() {
-    return SerializationUtils.serialize(_values);
+    int sizeInBytes = 0;
+    byte[][] cache = new byte[_values.length][];
+    for (int i = 0; i < _values.length; i++) {
+      Object value = _values[i];
+      if (value == null) {
+        continue;
+      }
+
+      if (value instanceof Integer) {
+        sizeInBytes += Integer.BYTES;
+      } else if (value instanceof Long) {
+        sizeInBytes += Long.BYTES;
+      } else if (value instanceof String) {
+        cache[i] = ((String) value).getBytes(StandardCharsets.UTF_8);
+        sizeInBytes += cache[i].length + Integer.BYTES;
+      } else if (value instanceof ByteArray) {
+        cache[i] = ((ByteArray) value).getBytes();
+        sizeInBytes += ((ByteArray) value).length() + Integer.BYTES;
+      } else if (value instanceof Float) {
+        sizeInBytes += Float.BYTES;
+      } else if (value instanceof Double) {
+        sizeInBytes += Double.BYTES;
+      } else if (value instanceof BigDecimal) {
+        cache[i] = BigDecimalUtils.serialize((BigDecimal) value);
+        sizeInBytes += cache[i].length + Integer.BYTES;
+      } else {
+        throw new IllegalStateException("Data type not supported for serializing Primary Key with value: " + value);
```

Comment:
We can handle `null` here:
```suggestion
        throw new IllegalStateException(String.format("Unsupported value: %s of type: %s", value, value != null ? value.getClass() : null));
```